cpp/src/arrow/scalar_test.cc

```diff
@@ -1135,6 +1136,25 @@ class TestListScalar : public ::testing::Test {
     ASSERT_NE(a0->hash(), b0->hash());
   }
 
+  void TestCast(ScalarType& scalar, const std::shared_ptr<DataType>& to_type) {
```

Comment:
Can we avoid non-const refs?
```suggestion
  void TestCast(const ScalarType& scalar, const std::shared_ptr<DataType>& to_type) {
```